signature/jws/envelope.go

```diff
@@ -0,0 +1,190 @@
+package jws
+
+import (
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"strings"
+
+	"github.com/golang-jwt/jwt/v4"
+	"github.com/notaryproject/notation-core-go/signature"
+	"github.com/notaryproject/notation-core-go/signature/internal/base"
+)
+
+const MediaTypeEnvelope = "application/jose+json"
+
+func init() {
+	if err := signature.RegisterEnvelopeType(MediaTypeEnvelope, NewEnvelope, ParseEnvelope); err != nil {
+		panic(err)
+	}
+}
+
+type envelope struct {
+	internalEnvelope *jwsEnvelope
+}
+
+func NewEnvelope() signature.Envelope {
+	return &base.Envelope{
+		Envelope: &envelope{},
+	}
+}
+
+func ParseEnvelope(envelopeBytes []byte) (signature.Envelope, error) {
+	var e jwsEnvelope
+	err := json.Unmarshal(envelopeBytes, &e)
+	if err != nil {
+		return nil, err
+	}
+	return &base.Envelope{
+		Envelope: &envelope{internalEnvelope: &e},
+		Raw:      envelopeBytes,
+	}, nil
+}
+
+func (e *envelope) Sign(req *signature.SignRequest) ([]byte, error) {
+	ks, err := req.Signer.KeySpec()
+	if err != nil {
+		return nil, &signature.MalformedSignRequestError{Msg: err.Error()}
+	}
+	alg := ks.SignatureAlgorithm()
+
+	signedAttrs, err := getSignedAttrs(req, alg)
+	if err != nil {
+		return nil, err
+	}
+
+	// parse payload as jwt claims
+	var payload jwt.MapClaims
+	if err := json.Unmarshal(req.Payload.Content, &payload); err != nil {
+		return nil, err
+	}
+
+	compact, err := sign(payload, signedAttrs, req.Signer)
+	if err != nil {
+		return nil, &signature.MalformedSignRequestError{Msg: err.Error()}
+	}
+
+	certs, err := req.Signer.CertificateChain()
+	e.internalEnvelope, err = generateJws(compact, req, certs)
+	if err != nil {
+		return nil, err
+	}
+
+	return json.Marshal(e.internalEnvelope)
+}
+
+func (e *envelope) Verify() (*signature.Payload, *signature.SignerInfo, error) {
+	if e.internalEnvelope == nil {
+		return nil, nil, &signature.SignatureNotFoundError{}
+	}
+
+	if len(e.internalEnvelope.Header.CertChain) == 0 {
+		return nil, nil, &signature.MalformedSignatureError{Msg: "malformed leaf certificate"}
+	}
+
+	cert, err := x509.ParseCertificate(e.internalEnvelope.Header.CertChain[0])
+	if err != nil {
+		return nil, nil, &signature.MalformedSignatureError{Msg: "malformed leaf certificate"}
+	}
+
+	// verify JWT
+	compact := strings.Join([]string{
+		e.internalEnvelope.Protected,
+		e.internalEnvelope.Payload,
+		e.internalEnvelope.Signature}, ".")
+	err = verifyJWT(compact, cert)
```

Comment:
nit: line 95 can be moved in the if statement.

Reply:
Sure.